### ISSUE TITLE
Awaiting a never-ending Future results in an early exit

### ISSUE THREAD

**Issue by pattobrien** [dart-lang/sdk#51462]
### Issue description

When running the following Dart program, the second print statement isn't reached at all, and instead the program exits with no errors or exceptions being thrown.

```dart
import 'dart:async';

// This future will never resolve
final neverEndingCompleter = Completer<bool>();

void main(List<String> args) async {
  print('start program');
  await neverEndingCompleter.future;
  print('program has completed'); // never prints to console
}

```

Console output:

```sh
Connecting to VM Service at http://127.0.0.1:57053/5LdbTbPh4mU=/
start program
Exited
```

No exceptions or errors are displayed along with the early exit (even when setting the debugger breakpoints for 'All Exceptions' and 'Uncaught Exceptions' to `true`), which means that any non-completing future will cause a silent failure that is extremely difficult to debug. As an example, in a personal project of mine it took 5+ hours to narrow this root cause down.

### Environment

Dart SDK version: 2.19.2 (stable) (Tue Feb 7 18:37:17 2023 +0000) on "macos_arm64"

```
[✓] Flutter (Channel stable, 3.7.3, on macOS 13.2.1 22D68 darwin-arm64, locale en-US)
    • Flutter version 3.7.3 on channel stable at /Users/pattobrien/fvm/versions/3.7.3
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 9944297138 (10 days ago), 2023-02-08 15:46:04 -0800
    • Engine revision 248290d6d5
    • Dart version 2.19.2
    • DevTools version 2.20.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc1)
    • Android SDK at /Users/pattobrien/Library/Android/sdk
    • Platform android-31, build-tools 33.0.0-rc1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.3

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] VS Code (version 1.75.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.58.0

[✓] Connected device (1 available)
    • macOS (desktop) • macos • darwin-arm64 • macOS 13.2.1 22D68 darwin-arm64

[✓] HTTP Host Availability
    • All required HTTP hosts are available
```

**Comment by lrhn** [comment]
Your asynchronous function does not return, meaning that the future it returns never completes.
It's impossible to say why without looking at the code.

The most common reason is that you introduce a zone with an error handler, then returns a future from inside that zone to outside of it, and the inner future completes with an error. That error gets blocked by the zone and never reaches the listeners outside of that zone.

**Comment by pattobrien** [comment]
Thanks so much for a quick reply, especially on the weekend!

I've identified zoned error handling in `package:riverpod`, and located an [issue](https://github.com/rrousselGit/riverpod/issues/1383) related to mishandling of zoned errors within `FutureProvider` - this is most likely the root cause. I'll confirm this at some point today and will close this issue accordingly.

Though this is working to spec, its very unfortunate that there's no way to use debug tools to  identify zoned errors (e.g. having breakpoint option 'All Exceptions for All Zones' or including zoned errors in 'All Exceptions', whichever makes more sense). I'm unsure of the feasibility or value of such functionality, I just can't imagine I'm the only one who has spent hours locating such exceptions.

What do you think? I can repurpose this issue around that suggestion or open a new issue, if you believe its worthwhile.

**Comment by pattobrien** [comment]
I've been able to fully reproduce the error when awaiting a future that never completes, see below for an example snippet:

```dart
import 'dart:async';

void main(List<String> args) async {
  print('start program');
  await globalCompleter.future;
  print('program has completed'); // never prints to console
}

final globalCompleter = Completer<bool>();
```

This seems like a clear bug, especially given the difficulty with debugging such a quiet failure. I'll add this info to the original issue description.

**Comment by mraleph** [comment]
Duplicate of https://github.com/dart-lang/sdk/issues/23797

**Comment by lrhn** [comment]
#23797 is one of the issues here. The other is https://github.com/dart-lang/sdk/issues/49457

I'd love to fix the latter, but it's non-trivial to do it without breaking existing code. It's not currently being worked on, but if we find a way, perhaps library versioning, I'd love to try again.
